lib/src/front_end/ast_node_visitor.dart

```diff
@@ -927,17 +930,88 @@ class AstNodeVisitor extends ThrowingAstVisitor<void>
 
   @override
   void visitSwitchExpression(SwitchExpression node) {
-    throw UnimplementedError();
+    var list = DelimitedListBuilder.switchBody(this);
+
+    createSwitchValue(node.switchKeyword, node.leftParenthesis, node.expression,
+        node.rightParenthesis);
+    writer.space();
+    list.leftBracket(node.leftBracket);
+
+    for (var member in node.cases) {
+      list.add(member);
+    }
+
+    list.rightBracket(node.rightBracket);
+    writer.push(list.build());
   }
 
   @override
   void visitSwitchExpressionCase(SwitchExpressionCase node) {
-    throw UnimplementedError();
+    if (node.guardedPattern.whenClause != null) throw UnimplementedError();
+
+    visit(node.guardedPattern.pattern);
+    writer.space();
+    finishAssignment(node.arrow, node.expression);
   }
 
   @override
   void visitSwitchStatement(SwitchStatement node) {
-    throw UnimplementedError();
+    createSwitchValue(node.switchKeyword, node.leftParenthesis, node.expression,
+        node.rightParenthesis);
+
+    // Attach the ` {` after the `(` in the [ListPiece] created by
```

Comment:
The first part made more sense when it followed a closing paren.

```suggestion
    // Attach the ` {` after the `)` in the [ListPiece] created by
```

Reply:
Oops, thanks for noticing that.